### PR TITLE
WIP: HTML5 Push Notifications

### LINKED ISSUE #10804: Notifications using Javascript Notification API
## Description
Current Gitea's notification system is kinda useless in my opinion. I get a lot of spam of notifications (new issues, closed issues, new issues comments, merged PRs (in a micro service project this becomes an huge spam of notifications), etc). When I actually see my notifications, there are usually dozens of them so I just click to mark them all as seen without actually seeing them.

Gitea should use [Notifications API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API) so I can see System Notifications while I use my IDE or something.

Gitea should also support the user to setup which events he wants to get notifications from on the settings page, ie I want to get notifictions only for new issues, comments on issues and comments on PRs.

Maybe gitea should use something like [Onesignal](https://onesignal.com/webpush) so it can send push notifications even if the user doesn't have any gitea's tab opened.

This issue is an improvement of #9327

### PR DESCRIPTION
**Pull request by jamesorlakin**
**Edit: This PR is currently being rewritten. I'll force push it when I'm done but feel free to see my `pushNotificationsNew` branch in the meantime.**

---

Fixes #10804, #9327.

This is a 'mostly there but unpolished' PR to implement push notifications using the Web Push and Notification API, powered by the webpush-go library. I feel this is a very useful piece of functionality to be productive, especially if emails aren't configured.

---

To give an overview:
1. Web Push servers need VAPID keys, a Base64'd asymmetric keypair. Gitea will generate a keypair on startup and add to app.ini if not set - similar to JWT secrets.
2. The user pushes a button to opt-in to notifications.
3. The JS requests the notification permission from the user. This permission also grants access to the service worker's PushManager to enable Web Push. _For Firefox at least it needs to be triggered by a user event to reduce spam._
4. If granted, the JS subscribes the service worker for Web Push using the VAPID public key embedded in the JS config object. This will then return a URL, auth token, and a signing key from the browser's vendor. (Further requests to subscribe will return the same details) _Currently this includes Chrome, Firefox, and Edge. Safari has its own proprietary system I haven't looked into._
5. The URL, auth token, and signing key are POSTed to Gitea under a new endpoint. The endpoint will send a test notification in order to verify the provided details before storing it against the user in the Web Push subscription table.
6. Done! :tada:

Whenever a notification model is created or updated we send a push notification:
1. Fetch all the subscribed pushes for a user.
2. Loop through each one and send a JSON object containing a message about the issue/PR and a URL.
3. If the HTTP code isn't 201 (created), delete the subscription. Under the spec 410 (gone) is sent for expired subscriptions but I think there's other codes if the auth/signing is wrong, etc. (**Edit: This is a bit dumb in retrospect. We should ignore 5xx errors as they shouldn't be our fault and are likely temporary.**)
4. The service worker receives the JSON object and uses the Notification API to show a corresponding notification.

---

As a result the service worker is required to be enabled - it's quite key! Having this lays the foundation for live messaging outside of notifications later on: updating the bell icon, informing the user of new activity while viewing a PR, etc.

This is WIP because I have some things outstanding/discussing:
- [ ] ~~Decouple the sending of pushes from the actual web request that triggered it. Would a simple goroutine be sufficient? Let's say a push service took 10s... ouch!~~
- [ ] ~~Refactor web push to be a module using the notification service entrypoint, rather than hijacking events on the notification model~~
- [ ] **Figure out the Notifier queue system and use that instead of hijacking the notification model**
- [ ] Localisation. I can do most of the server-side bits but not sure about the JS button. It won't show until it passes feature discovery. I guess I should put it in the template and add all required localised messages as custom attributes.
- [x] Add the repo name to the notification - oops!
- [x] Only remove subscription when encountering a 4xx error.
- [ ] Rename `.ini` config names
- [ ] More error handling on the JS side
- [ ] Figure out a nicer/more obvious UI to opt-in.
- [ ] Allow the user to opt-out
- [ ] Testing?!

And as always, even minor nits are welcome! A lot of the logic is _in_ the models, but I need it to automatically delete old hooks and react to notifications. It was a module but I caused an import cycle... 🤦‍♂ 

---

image
Gitea notifications new issue

### CODE REVIEW

**Comment by SpaWn2KiLl** [comment]
Wow! 👏
@6543 @zeripath come have a look at this!

**Comment by 6543** [comment]
Can you resolve the conflict - Ill look at it afterwards :)

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 1]
```diff
@@ -131,7 +131,7 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `REACTIONS`: All available reactions. Allow users react with different emoji's.
 - `DEFAULT_SHOW_FULL_NAME`: **false**: Whether the full name of the users should be shown where possible. If the full name isn't set, the username will be used.
 - `SEARCH_REPO_DESCRIPTION`: **true**: Whether to search within description at repository search on explore page.
-- `USE_SERVICE_WORKER`: **true**: Whether to enable a Service Worker to cache frontend assets.
+- `USE_SERVICE_WORKER`: **true**: Whether to enable a Service Worker to cache frontend assets and enable push notifications on supported browsers.
```
Comment (silverwind) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md:134]: Don't piggy-back on the service-worker pref, these are two distinct features. Add a new one, like `USE_PUSH_NOTIFICATIONS`.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 2]
```diff
@@ -289,6 +289,8 @@ set name for unique queues. Individual queues will default to
 
 - `INSTALL_LOCK`: **false**: Disallow access to the install page.
 - `SECRET_KEY`: **\<random at every install\>**: Global secret key. This should be changed.
+- `WEB_PUSH_PUBLIC_KEY`: **\<random at every install\>**: VAPID key pair used for Web Push notifications
+- `WEB_PUSH_PRIVATE_KEY`: **\<random at every install\>**: VAPID key pair used for Web Push notifications
```
Comment (silverwind) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md:293]: `PUSH_NOTIFICATIONS_PUBLIC_KEY` and `PUSH_NOTIFICATIONS_PRIVATE_KEY` would be more better imho.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md (hunk 1 above)
Comment (jamesorlakin) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md:134]: True - I'll pop a note saying the service worker is required as well.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md (hunk 2 above)
Comment (jamesorlakin) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md:293]: In the codebase I've tried to be consistent with 'Web Push', but I could change the user-facing config quite easily

**Comment by jamesorlakin** [comment]
Go's modules confuse me a fair bit!

So CI is failing because my machine adds `go 1.13` to the vendored package's `go.mod` and CI seemingly doesn't:
image

I'm using Go 1.14.1 locally, but if I try and do `make vendor` with 1.13 the vendored `modules.txt` goes ...haywire:
image

I'm using Windows if that makes any difference.

**Comment by 6543** [comment]
@jamesorlakin go 1.14 is requred if you exec `make vendor`

**Comment by 6543** [comment]
you should also format the code with `make fmt` and commit the result

**Comment by jamesorlakin** [comment]
Cheers, but it's insistnent on putting the `go 1.13` back into that file even with `make fmt && make vendor` on Go 1.14.1 for me. 😕

**Comment by 6543** [comment]
> Cheers, but it's insistent on putting the `go 1.13` back into that file even with `make fmt && make vendor` on Go 1.14.1 for me.

this could be indeed a windows only issue

**Comment by 6543** [comment]
Pleace resolve conflicts

**Comment by silverwind** [comment]
serviceworker implementation has changed significantly since this PR as it's now handled by workbox so this will need quite a few updates.

**Comment by 6543** [comment]
@jamesorlakin IE support was droped in v1.13

**Comment by jamesorlakin** [comment]
Even better 👍

**Comment by silverwind** [comment]
I imagine the new implementation can re-use the `EventSource` to add a new `notification` event and then push that to the serviceworker for display. These parts don't depend on workbox, but they do have [this example](https://developers.google.com/web/tools/workbox/guides/codelabs/webpack#inject) which includes push notifications. Also, I wonder if desktop browsers would need a fallback to the older web notifications API.

**Comment by mohe2015** [comment]
@silverwind
> Also, I wonder if desktop browsers would need a fallback to the older web notifications API.

I'm pretty sure they don't as I have implemented push notifications for desktop and mobile a while ago using the standard APIs.

**Comment by silverwind** [comment]
I guess fallback to web notification is a requirement because there are scenarios where serviceworker will be disabled (either by configuraiton or when the user is in private browsing mode).

**Comment by mohe2015** [comment]
@silverwind 
> I guess fallback to web notification is a requirement because there are scenarios where serviceworker will be disabled 
There is still the default notification system (email, just the bell in the top right of the web interface)
Of course people may still want notifications but when using Web Notifications they need to keep their browser open and you would need to poll or use something like websockets to get updates.

> (either by configuration
I think then a fallback is not intended either (or do you mean disabled in the browser?)

> or when the user is in private browsing mode (which may disable serviceworker in some circumstances)).
I think then you can't even enable the push notifications and I don't think this is the way most people who want push notifications use their browser.

> Maybe there is a third-party module that can handle such a fallback.
Unfortunately I don't think so as the technology is quite different (Web Notifications is just showing a Notification and Push Notification is also getting updates to the browser)

So in my personal opinion I would see Push Notifications as a progressive enhancement so I don't think a fallback is required. This would also simplify the implementation.

**Comment by silverwind** [comment]
> you would need to poll or use something like websockets to get updates.

We do have SSE/Eventsource implemented for server-side push things. Currently it's only used to update the notification count. We probably will want to switch this to websockets because SSE has some quirks and limitations.

What I mean is that web notifications would still be possible even without a service worker available while push notifications require it, but I guess we don't need to handle it as it would complicate the implementation (e.g. need to handle multiple tabs racing to push the same notification etc).

**Comment by jamesorlakin** [comment]
> We do have SSE/Eventsource implemented for server-side push things.

Correct me if I'm wrong here, but I thought that was handled by an AJAX timer?

I think for now we should stick to feature discovery (does the browser support SWs, push, etc) and only show the option if the service worker is enabled in the config. Still showing notifications outside of notifications is a nice idea, but we need to make it clear this fallback mode would not provide notifications if no Gitea tab is open.

**Comment by silverwind** [comment]
See https://github.com/go-gitea/gitea/blob/master/web_src/js/features/notification.js#L30, it's using EventSource and falls back to ajax, at least that's the plan.

> I think for now we should stick to feature discovery (does the browser support SWs, push, etc) and only show the option if the service worker is enabled in the config

I would make a second option and only enable push notifications if both are enabled, that way one can still benefit from SW caching while keeping notifications off if desired.

**Comment by catdevnull** [comment]
Is this ever gonna land? :#

**Pull request closed without merging**